templates/events/macros/event_format.html

```diff
@@ -0,0 +1,95 @@
+{% if resource.extra.location -%}
+  {% if icon %}{{ shortcodes::icon(icon="ph-map-pin") }}{% endif -%}
+  {% if link and resource.extra.location_url -%}
+    <a target="_blank" rel="noopener" href="{{ resource.extra.location_url }}" aria-label="Location for {{ resource.title }}">{{ resource.extra.location }}</a>
+  {% elif not link or resource.components is containing("archive") -%} 
+    {{ resource.extra.location }}
+  {% else -%} 
+    {{ rooms::room_link(name=resource.extra.location) }}
+  {% endif -%}
+{% endif %}
+{%- endmacro -%}
+
+
+{%- macro event_time(e) -%}
+{%- if e.extra.display_time -%}
+  {{ e.extra.display_time }}
+{%- else -%}
+  {{ formatters::time(time=e.date) }}
+{%- endif -%}
+{%- endmacro event_time -%}
+
+
+{%- macro event_time_range(e) -%}
+{{ event_format::event_time(e=e) }}
+{%- if e.extra.end_time -%}   
+  {%- set parsed = e.extra.end_time | parse_date(base=e.date) -%}
+  -{{ formatters::time(time=parsed) }}
+{%- endif -%}
+{%- endmacro event_time -%}
+
+
+{%- macro event_date_range(e, day=false) -%}
+{%- if e.extra.end_time -%}{%- set parsed_end = e.extra.end_time | parse_date(base=e.date) -%}{%- endif -%}
+{%- set diff_date = not e.extra.display_date and parsed_end and e.date | date(format="%F") != parsed_end | date(format="%F") -%}
+{%- set diff_time = not e.extra.display_time and parsed_end and e.date | date(format="%T") != parsed_end | date(format="%T") -%}
+
+{%- set start_date = event_format::event_start_date(e=e) -%}
+{%- set end_date = event_format::event_end_date(e=e) -%}
+
+{# Start of date range, use display time if present #}
+{%- if e.extra.display_time -%}
+  {{ e.extra.display_time }}
+{%- else -%}
+  {{ formatters::time(time=start_date) }}
+{%- endif -%}
+{%- if diff_date -%}
+  {# Output start date if end_date is different #}
+  {{ formatters::date(date=start_date, day=day) }}
+{%- endif -%}
+
+{# Output dash if necessary #}
+{%- if diff_time or diff_date -%}{{ " - " }}{%- endif -%}
+{# Output end time & date if different #}
+{%- if diff_time -%}
+  {{ formatters::time(time=end_date) }}
+{%- endif -%}
+{{ " " }}
+{%- if diff_date -%}
+  {{ formatters::date(date=end_date, day=day) }}
+{%- elif e.extra.display_date -%}
+  {{ e.extra.display_date }}
+{%- else -%}
+  {{ formatters::date(date=start_date, day=day) }}
+{%- endif -%}
+{%- endmacro event_time -%}
+
+
+{% macro event_start_date(e, week_base="") -%}
+{% set start = e.date -%}
+{% if week_base -%}
+  {% set start = start | date(format="%A %T") -%}
+  {{ start | parse_date(base=week_base) -}}
+{% else -%}
+  {{ start -}}
+{% endif -%}
+{% endmacro event_time -%}
+
+
+{% macro event_end_date(e, week_base="") -%}
+
+{% set start = e.date -%}
+{% if week_base -%}
+  {% set start = start | date(format="%A %T") -%}
+  {% set start = start | parse_date(base=week_base) -%}
+{% endif -%}
+
+{%- if e.extra.end_time -%}
+  {{ e.extra.end_time | parse_date(base=start) -}}
+{% else -%} 
+{{ "1 hour" | parse_date(base=start) -}}
```

Comment:
i'd say dont do any diff time if unkown

Reply:
Discord needs an end time, and i think ical (strictly) does too. It's better than breaking stuff